### ISSUE TITLE
Message for Notifying about Hooks/Non-Tests

### ISSUE THREAD

**Issue by thomhurst** [microsoft/testfx#4087]
## Summary

A new message that can publish information about hooks/Non-Tests

## Background and Motivation

Currently if a hook fails, I have two options:
- attach the failure to an existing test node
- create a dummy test node to represent the hook

Those problems with these are, respectively:
- this implies the test failed which isn't the case, and it also complicates code bases more as the test object must be kept around for longer, and also timings. Do I count hook timings in the test that triggered it? Ideally no, it would have its own object with it's own timing.
- dummy test nodes are reported as new tests, and would then show under the test counters at the end, and IDEs would show them as runnable tests.

## Proposed Feature

A new message to represent hooks would be nice.

It would simplify my logic as I wouldn't have to keep test objects around for longer and delay their result publishing until hooks had finished.

Optionally, supplying a property of test IDs to them could represent what tests they relate to/trigger from. These could be used to drive IDE uis if they wanted to build something to show hook execution.

**Comment by nohwnd** [comment]
Do you please have some real-life examples where you would need this so I can better understand the ask? If the hook events are not tied to any test, would they be tied to some place in the test hierarchy, or just to the whole run?

**Comment by thomhurst** [comment]
I guess in a test explorer view, tests are usually grouped by Assembly > Class > Test anyway, so an entry could be added underneath the respected level they're hooking into.

But this new message type shouldn't give the option to be able to run explicitly. Just shown form informative purposes.

**Comment by nohwnd** [comment]
I am actually not sure if there is some message for this but there should be.

To me those hooks are not super special, it is test explorer that limits us here, and we do have a special mode for MSTest that adds dummy test entries for `*Initialize` and `*Cleanup`  to be able to represent them in test explorer. 

A more MTP oriented solution would be if the IDE understood the relations betwween tests, and then we would simply attach the results to the container that they belong to.

test cleanup error -> test "container"

class cleanup error -> class "container" that contains the tests (or maybe groups of tests and their retries)

assembly cleanup error -> assembly  which is parent of all the below,

and even further: single run , run with retries,  distributed per computer run, overall run. it scales as far as you need.

This was the intent of having the tree of tests rather than flat list of tests that have some metadata on them. Problem is going further with this idea when users would not be able to see half of the additional data in Test  Explorer.

**Comment by nohwnd** [comment]
**Needs**: research on what we have already in place, if anything, and how we could represent this in the MTP, and how we could then show it in TE or VSCode. 

The case where we have class or assembly cleanup errors are real asks from customers, but this does not have to be used only for errors.

**Comment by Evangelink** [comment]
Nice summary @nohwnd!

**Comment by nohwnd** [comment]
Similar request in #5248

**Comment by bradwilson** [comment]
For xUnit.net, our hierarchy of places where we can catch and need to report exceptions includes:

- Assembly-level
- Test collection-level
- Test class-level
- Test method-level
- Test case-level
- Test-level

Right now, the mapping from xUnit.net => MTP is purely based on test case ID, which ends up grouping multiple tests under a single test case when such conditions exist (non-pre-enumerated theory data being the primary one).

The only way I can make this work today without silently ignoring failures (which is what I'm doing today, incorrectly) is to track every test we run, associated with every level that it could be at, so that when one of these failures happens, I can add it as an 'extra failure' to every associated test.

The end result of this is that we must report multiple failures depending on where the failure happens. A test-level failure will just be a 1:1 mapping, but a test class failure (for example, a user throwing in the `Dispose` of a class fixture) will end up reporting every test in that test class with an additional failure result:

![Image](https://github.com/user-attachments/assets/196160e0-3a57-4bc9-b34c-754b9fa031f1)

Until Test Explorer can express arbitrary hierarchy that's controlled by the test framework (so I can associate the error with the correct place in the hierarchy), this forced replication is the best I can do.

**Comment by Evangelink** [comment]
@bradwilson @thomhurst assuming there is no technical limitation to what we can do with Test Explorer, what would be the ideal rendering for you? Some questions below to help drive the thouhts:
- do these messages need some association with test tree?
- do they need to be directly visible or could they be on a different part/UI

**Comment by thomhurst** [comment]
I think associating it with the test tree and making them visible would be useful.

With the existing test tree being like:

- Assembly
  - Class 1
    - Test 1
    - Test 2
  - Class 2
    - Test 1
    - Test 2

I think it'd be nice to have:

- Assembly
  - Before: SetupMethodName
  - Class 1
    - Before: SetupMethodName
    - Test 1
    - Test 2
    - After: CleanupMethodName
  - Class 2
    - Test 1
    - Test 2
  - After: CleanupMethodName

For the user it's clear what's run. And if a hook failed, we could report the failure on that node and not the tests, so it's simple and quick for the user to know where to look for fixing the failure.

**Comment by bradwilson** [comment]
> do these messages need some association with test tree?

I have two different thoughts here.

One is that a separate node for "non-test failures" could dynamically show up, and these things could end up there. From a textual perspective, we ensure that we describe the type of failure (for example, "Test Collection Cleanup Failure"), so grouping them all into one place is a reasonable first step. This wouldn't even require support for arbitrary hierarchy, since we'd just report these as a new type of node (not a test node).

The other is if/when test frameworks can define arbitrary hierarchy that's represented by Test Explorer. So in this scenario, a "fully populated" tree for an xUnit.net test would look like:

    Test Assembly
    |
    -- Test Collection
       |
       -- Test Class
          |
          -- Test Method
             |
             -- Test Case
                |
                -- Test

Today in the API, our only reporting node lies at the "test case" level. For `[Fact]` tests, and for `[Theory]` tests where data is pre-enumerated, there is a 1:1 mapping between Test Case and Test. For `[Theory]` tests where data is not/cannot be pre-enumerated, this is a 1:* mapping between Test Case and Test (one per data row). For third party test framework extensions, there are no rules other than the fact that we support 1:* mappings between Test Case and Test. One simple example: in our `[CulturedFact]` and `[CulturedTheory]` for our own tests, we do 1:* mappings between Test Case and Test, one for each specified culture (defaults to both `en-US` and `fr-FR`). Due to the support for theories, it can actually be a 1:(m*n) due to "non-prenumerated data x cultures".

If all of these things were groupable in the UI, then you could imagine being able to designate a failure node as belonging to a specific place in the hierarchy ("this Test", "test Test Case", "this Test Collection", etc.) If/when the user decides to collapse these things in the UI, then you'd have to decide where to place these based on moving up the tree to the next visible node (i.e., all "Test Collection Cleanup Failure" messages would land in the appropriate Test Assembly if the user is not using "Test Collection" as a grouping mechanism.).

> do they need to be directly visible or could they be on a different part/UI

I'm not sure what this means, but I will say that the user requirements be that (a) the test run fails, and (b) it be easy to understand why it failed, when it didn't fail because of a test.

This means, if you bury these failures in a hard to see/hard to find place, users will be upset (with me, of course, because these issues always get blamed on the framework). It would be monumentally worse if you let them believe the tests passed when there were cleanup failures.

**Comment by bradwilson** [comment]
Also, it goes without saying that these non-test nodes are not "runnable".

**Comment by bradwilson** [comment]
Let me clarify on that a bit.

You would obviously not see these cleanup failures during discovery, and they would by definition not be runnable, because they aren't tests.

However, in a world where you could imagine that arbitrary hierarchy is supported, then during discovery we should absolutely be able to say a "Test Case" has a parent `TestNode` that is the "Test Method", and the user could ask to run the "Test Method", which would in turn run all the tests that are children underneath it. Any cleanup errors reported at the "Test Method" level would themselves not be runnable, so they are not strictly speaking a `TestNode`. However, the user could attempt to re-run/debug the "Test Method" in an effort to isolate/reproduce the cleanup failure without having to re-run the whole test assembly (which is where we would be without support for arbitrary hierarchy).

So to me the support for hierarchy is more than just about UI & grouping; it also represents a framework-defined list of "groups of runnable things".

**Comment by bradwilson** [comment]
It's also worth noting that "Test Class" and "Test Method" are optional things in our hierarchy. You may end up with "Test Case"s that are directly parented in "Test Collection"s. This is allow for test definitions that are not based on the strongly-CLR-aligned concepts of class and method.

**Comment by thomhurst** [comment]
I agree with the custom hierarchy being the ideal outcome.

Runnable nodes that I envision are (and some even are currently by the test explorer but it must use it's own logic as to how, instead of asking the framework):

Test Suite (everything discovered)
Assembly
Class
Test Group (data driven with different cases)
Test case (a single test case, data or non data driven)

Also there could be the option of having multiple parents. This could be complicated to show in the UI. Maybe a filter selecting on how they are displayed?

Because as well as the above, I could see grouping by 'Category', or some custom properties.

Each of these nodes could define their children, so the test explorer doesn't have to perform any custom logic on what to run. It can just pass a test node id filter containing the IDs of all the descendants, which are properties on each node.

A child could also be this new non-runnable node, which would represent these hooks.

Then everything is defined by the framework, as it should be imo.

**Comment by bradwilson** [comment]
> Also there could be the option of having multiple parents.

This is not modeled in the API today, and frankly it becomes incredibly complicated. Is there no "best" choice by default?

**Comment by bradwilson** [comment]
> Because as well as the above, I could see grouping by 'Category', or some custom properties.

I believe this is better served by filtering than grouping, personally.

**Comment by thomhurst** [comment]
Maybe complicated for any sort of UI test runner, yes. For the frameworks it'd be simple though. Just define a node with children. Running any node will run all of it's descendants. And then any hierarchy you've designed is easily runnable without having to conform to any specific structure.

**Comment by bradwilson** [comment]
> Maybe complicated for any sort of UI test runner, yes.

Right, that's what I meant. Sorry. Confusing because I lumped it in with the fact that the API doesn't model this (presumably because they were considering UI).

**Comment by Evangelink** [comment]
Thanks guys, so far this is totally aligned with what we had in mind and wanted to design but we're not able to provide due to lack of investments on Test Explorer. I'll double check if the ticket we had is internal only or public (in which case I'll link it here).

Because of these limitations, we did a really hacky solution (still needs way more work) on MSTest similar to what you reported above @bradwilson .

As for the different kinds of grouping I think that's a "UI problem" that is "easily" sortable (with some discussions). Assuming the framework reports any arbitrary tree, you then have the following options: keep user ordering and hide (or transform as non-runnable nodes) excluded items - this is allowing to respect framework grouping and falls under the filtering described by Brad OR change grouping based on the thing selected. Potentially both make sense and could be different UI features (grouping and filtering).
